// nylas/lib/models/calendar-restful-model-collection.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var calendar_1 = __importDefault(require("./calendar"));
var restful_model_collection_1 = __importDefault(require("./restful-model-collection"));
var CalendarRestfulModelCollection = /** @class */ (function (_super) {
    __extends(CalendarRestfulModelCollection, _super);
    function CalendarRestfulModelCollection(connection) {
        var _this = _super.call(this, calendar_1.default, connection) || this;
        _this.connection = connection;
        _this.modelClass = calendar_1.default;
        return _this;
    }
    CalendarRestfulModelCollection.prototype.freeBusy = function (options, callback) {
        return this.connection
            .request({
            method: 'POST',
            path: "/calendars/free-busy",
            body: {
                start_time: options.startTime || options.start_time,
                end_time: options.endTime || options.end_time,
                emails: options.emails
            }
        })
            .then(function (json) {
            if (callback) {
                callback(null, json);
            }
            return Promise.resolve(json);
        })
            .catch(function (err) {
            if (callback) {
                callback(err);
            }
            return Promise.reject(err);
        });
    };
    return CalendarRestfulModelCollection;
}(restful_model_collection_1.default));
exports.default = CalendarRestfulModelCollection;
